src/util/json_stream.h

```diff
@@ -0,0 +1,174 @@
+/*******************************************************************\
+
+Module:
+
+Author: Peter Schrammel
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_JSON_STREAM_H
+#define CPROVER_UTIL_JSON_STREAM_H
+
+#include <memory>
+#include <ostream>
+
+#include "json.h"
+#include "invariant.h"
+
+class json_stream_objectt;
+class json_stream_arrayt;
+
+/// This class provides a facility for streaming JSON objects directly to the
+/// output instead of waiting for the object to be fully formed in memory and
+/// then print it (as done using `jsont`). This reduces memory consumption
+/// when preparing e.g. large traces for output.
+///
+/// `json_streamt` is the base class from which the classes for streaming
+/// arrays (`json_stream_arrayt`) and objects (`json_stream_objectt`) are
+/// derived. `json_streamt` wraps an output stream and stores the current
+/// element to be output. This can be either a child stream (of type
+/// `json_streamt` that outputs JSON objects incrementally) or a
+/// non-streaming JSON `object` (of type `jsont` that will be output as a
+/// whole) To ensure consistency of the output, a class invariant is that
+/// there is at most one child stream at any time. For this reason, the
+/// existing child stream is flushed and closed when creating a new child
+/// stream.
+class json_streamt
+{
+public:
+  /// Outputs the current current child stream and closes this JSON stream
+  void close()
+  {
+    if(open)
+    {
+      output_child_stream();
+      output_finalizer();
+      open = false;
```

Comment:
Wouldn't it be good enough to use the `child_stream` as an indicator (resetting it after it is closed)?